include/seastar/http/request.hh

```diff
@@ -109,15 +113,84 @@ struct request {
 
     /**
      * Search for the last query parameter of a given key
-     * @param key the query paramerter key
+     * @param key the query parameter key
      * @return the query parameter value, if it exists or empty string
      */
     sstring get_query_param(const sstring& key) const {
-        auto res = query_parameters.find(key);
-        if (res == query_parameters.end()) {
-            return "";
+        if(_query_params.empty()) {
+            auto res = query_parameters.find(key);
+            if (res != query_parameters.end()) {
+                return res->second;
+            }
+        } else {
+            auto res = _query_params.find(key);
+            if (res != _query_params.end()) {
+                return res->second.back();
+            }
         }
-        return res->second;
+
+        return "";
+    }
+
+    /**
+     * Search for all query parameters of a given key
+     * @param key the query parameter key
+     * @return a vector of all query parameter values, if it exists or an empty vector
+     */
+    const std::vector<sstring>& get_query_param_array(std::string_view key) const {
+        if(auto res = _query_params.find(key); res != _query_params.end()) {
+            return res->second;
+        }
+        static const std::vector<sstring> empty_vector;
+        return empty_vector;
+    }
+
+    /**
+     * Get all query parameters
+     * @return a map of all query parameters
+     */
+    const query_param_type& get_query_params() const {
+        return _query_params;
+    }
+
+    /**
+     * Set a query parameter value
+     * @param key the query parameter key
+     * @param value the query parameter value
+     * @return a reference to this request object
+     */
+    request& set_query_param(std::string_view key, std::string_view value) {
+        return set_query_param(key, {value});
+    }
+
+    /**
+     * Set a query parameter value
+     * @param key the query parameter key
+     * @param values the query parameter values
+     * @return a reference to this request object
+    */
+    request& set_query_param(std::string_view key, std::initializer_list<std::string_view> values) {
+        _query_params[sstring(key)] = std::vector<sstring>(values.begin(), values.end());
+        return *this;
+    }
+
+    /**
+     * Set a query parameter value
+     * @param key the query parameter key
+     * @param values the query parameter values
+     * @return a reference to this request object
+    */
+    request& set_query_param(std::string_view key, const std::vector<sstring>& values) {
+        _query_params[sstring(key)] = values;
```

Comment:
nic: This copies thev ector. If the 2nd argument is a value and the assignment uses std::move() that'd be slightly better API